delete.go

```diff
@@ -66,6 +66,14 @@ status of "ubuntu01" as "stopped" the following will delete resources held for
 			}
 			return err
 		}
+		// When --force is given, we kill all container processes and
+		// then destroy the container. This is done even for a stopped
+		// container, because (in case it does not have its own PID
+		// namespace) there may be some leftover processes in the
+		// container's cgroup.
+		if force {
```

Comment:
I don't know whether `containerd` always uses `runc kill` and `runc delete -f` or not when the container is in `stopped` state.

Reply:
I guess it is, as without `runc delete` there will be some leftovers. Note you don't have to call `runc kill` before `runc delete -f`.

Also, if the processes are not properly killed, it will be impossible to start a new container with the same cgroup (thanks to #3223 and #3782).

Reply:
https://github.com/opencontainers/runc/pull/4048#discussion_r1382527177

Reply:
I think this PR didn't fix #4047 .

Reply:
We should have a test case for #4047 , I'll add one once I have a time, and if anybody can add this test case, it will be appreciated.

Reply:
Maybe we should need this commit:
https://github.com/opencontainers/runc/pull/4048/commits/48ff07c2ee1c375116ec34c5a648d53d60f83b79

Reply:
> I'll add one once I have a time, and if anybody can add this test case, it will be appreciated.

The testcase: https://github.com/opencontainers/runc/pull/4048/commits/cc9addd8d87d635b0546cb6cf3cb50a229df942d

I think we should have the ability to delete the container like before, or tell the user why you can't delete the container , it's because there are still some processes in the cgroup even the container has stopped.
I can't accept we have deleted all state dir, but leave the cgroup and some container processes alive, when the user call `runc delete` without `-f`.

Reply:
@lifubang Ah, you mean `runc delete` case. It is not listed in #4047.

Your test case is basically the same test as added by this PR, except for using `runc delete` rather than `runc delete -f`?

What do you think the correct behavior should be? I think, runc should error out, saying something like "container cgroup is not empty (pids: XXXX YYYY ZZZZ), use runc kill first".

Reply:
> It is not listed in #4047.

Sorry, it’s a bug after I opened this issue.


> What do you think the correct behavior should be?

It’s a problem.
I like error it out, but it is not compatible of previous version.
Another think, if we need to always run ‘kill’ before ‘delete’, why we need to have ‘-f’, maybe we can drop it.

Reply:
Sometimes, when it makes sense, and we don't expect major breakage or have a workaround, we do introduce incompatible changes. For example, it was possible to run two containers in the same cgroup -- not it's not. It was possible to do `runc exec` in a paused container -- now it's not. These changes are for good and to prevent other bugs down the line.

Reply:
> Another think, if we need to always run ‘kill’ before ‘delete’, why we need to have ‘-f’, maybe we can drop it.

Well, I think that
1. Most containers have own private PID ns, what we discuss here is a corner case.
2. `runc delete -f` is a shortcut for `runc kill $CTID KILL && runc delete $CTID` which is handy, and unlike `kill -a` we can't drop this option because it changes the behavior (while `runc kill` no longer needs `-a`, it works fine without it.

Reply:
Also left a comment in https://github.com/opencontainers/runc/issues/4047#issuecomment-1797099129